### ISSUE TITLE
some issues

### ISSUE THREAD

**Issue by xXuHaiyang** [wilile26811249/Next-ViT#1]
Hello! Nice work with some slight drawbacks.
1. In 3.5 "The head dim in E-MHSA and MHCA is set as 32". 
    But in MHCA, actually you set groups =32 but not groups = dim // 32.
    And in E-MHSA, you set inner dim to be 64(maybe you could check the overall operation and dim)
2. In 3.5 "spatial reduction ratios in E-MHSA is [8, 4, 2, 1] in different stages"
    But you set a uniform sr_ratio, or "stride" in your code
3. In 3.5 "The expansion ratios of MLP layer are set as 3 for NCB and 2 for NTB, respectively."
    You simply ignore this. And this will cause a dramatic drop in parameters and FLOPs, as well as the model performance.
4. With respect to Block, I think here is a bug, because the author didn't mentioned it detailed so we can't determined the input and output dim between "L" blocks. I think it proper to implement this as you did, i.e, in_dim for "L-1" and turn to out_dim at last repeat.

**Comment by wilile26811249** [comment]
Thank you for your advices! 

1. Yes, should set the groups to dim//32 not set to 32.
About the E-MHSA block, do you have any suggestion?

2. Yes, I missed the "spatial reduction ratio" part.
And I also found an issue, the paper says the "spatial reduction ratio" in stage1 should be set to 8, but in stage1 it doesn't contain the E-MHSA block, maybe I miss something.

3. Yes, I missed the "expansion ratio" part. Thank you.

**Comment by xXuHaiyang** [comment]
Not at all.
1. maybe "assert dim == self.inner_dim * self.heads" ? I'm not sure though.  If so, then init in NTB should be like 
    self.e_mhsa = E_MHSA(dim=first_part_dim, inner_dim=32, heads=first_part_dim//32, 
                                             stride=sr_ratio, drop=drop, drop_path=drop_path)

2. I'm also confused ...

This paper fused me. I've improve and complete your implementation and try to run it, everything is set according to the author's experimental settings. But the FLOPs is much lower than the author reports and train loss will be NAN in no more than 5 epochs. Have you tried to run it yet?

**Comment by wilile26811249** [comment]
No, I didn't run the NextViT on any dataset, just check the forward pass is correct...

**Comment by LightToYang** [comment]
> Not at all.
> 
> 1. maybe "assert dim == self.inner_dim * self.heads" ? I'm not sure though.  If so, then init in NTB should be like
>    self.e_mhsa = E_MHSA(dim=first_part_dim, inner_dim=32, heads=first_part_dim//32,
>    stride=sr_ratio, drop=drop, drop_path=drop_path)
> 2. I'm also confused ...
> 
> This paper fused me. I've improve and complete your implementation and try to run it, everything is set according to the author's experimental settings. But the FLOPs is much lower than the author reports and train loss will be NAN in no more than 5 epochs. Have you tried to run it yet?

low FLOPs may be caused by redundant downsampling in self.stage1
```
self.stage1 = nn.Sequential(
    # PatchEmbed(64, self.next_vit_channel[0]),
    nn.Conv2d(64, self.next_vit_channel[0], 1, 1, 0),
    Block(self.next_vit_channel[0], self.next_vit_channel[0], 1, 0, 1, 8, drop_path),
)
```

**Comment by Jiashi-Li** [comment]
Hi，thanks for your interest about Next-ViT, the officail repo have been open in https://github.com/bytedance/Next-ViT.

**Comment by xXuHaiyang** [comment]
> Hi，thanks for your interest about Next-ViT, the officail repo have been open in https://github.com/bytedance/Next-ViT.

Thanks for informing!

**Comment by wilile26811249** [comment]
> Hi，thanks for your interest about Next-ViT, the officail repo have been open in https://github.com/bytedance/Next-ViT.

Thanks for your informing!
